### DISCUSSION TITLE
Match anything between '#{' and '#}'

### DISCUSSION THREAD

**Discussion by FoamScience** [tree-sitter/tree-sitter#1480] [Q&A] [answered]
I'm writing a grammar for a new file-format/language, which has "code fences" of some sort:
```
code
#{
    /*
        Comment 
    */
    #include<vector>
    #include "someFile.H"

    { os << vector(-::sin($angle), 0, ::cos($angle)); }
#};
```

What I want to do is to skip parsing what's between "#{" and "#}" .
This looks like C-like comments but I don't have to worry about nested "#{" (but there will be nested "{}" blocks).
The best I could do is to:
```
    code: $ => seq(
        '#{',
        field("codeBody", token(
                    /[^#][^}]+/
        )),
        '}'
    ),
```
> Note how "#" from "#}" is parsed as part of the body
Of course, this fails for nested "{}" blocks; I feel like I absolutely need `lookaheads` for this to work but these are not supported; any ideas?

Thanks in advance

**Comment by ahelwer** [comment]
I have something similar in my language, nested blocks delimited by `(*` and `*)`. The only solution I found was to implement this in an external scanner. You can see the code [here](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/26bbaf5d07356cacbf21defa9316988bc732c8b3/src/scanner.cc#L244).

**Reply by FoamScience** [reply]
Thanks! Will try to implement it in a similar way then.

**Reply by FoamScience** [reply]
Can I have some help with this please (I'm awful with C cases); Here's what I've got:
```
  bool scan_block_comment_text(TSLexer* const lexer) {
    uint32_t nest_level = 0;
    CLexState state = CLexState_CONSUME;
    START_LEXER();
    eof = !has_next(lexer);
    switch (state) {
      case CLexState_CONSUME:
        if (eof) ADVANCE(CLexState_END_OF_FILE);
        if ('#' == lookahead) ADVANCE(CLexState_ASTERISK);
        ADVANCE(CLexState_CONSUME);
        END_STATE();
      case CLexState_ASTERISK:
        if ('}' == lookahead) ADVANCE(CLexState_END_OF_FILE);
        if ('{' == lookahead) ADVANCE(CLexState_L_PAREN);
        ADVANCE(CLexState_CONSUME);
        END_STATE();
      case CLexState_L_PAREN:
        ADVANCE(CLexState_CONSUME);
        END_STATE();
      case CLexState_END_OF_FILE:
        ACCEPT_TOKEN(CODE_BODY);
        END_STATE();
      default:
        END_STATE();
    }
  }
```
As I don't need to count for nesting, this should work, but no; stalls!

**Reply by ahelwer** [reply]
Stalling sounds like an infinite loop. What does the `tree-sitter parse -d` output look like?

**Comment by FoamScience** [comment]
`parse -d` goes on about:
```
process version:0, version_count:1, state:3, row:8, col:0
reduce sym:source_file_repeat1, child_count:2
shift state:3
process version:0, version_count:1, state:3, row:8, col:0
reduce sym:source_file_repeat1, child_count:2
shift state:3
process version:0, version_count:1, state:3, row:8, col:0
reduce sym:source_file_repeat1, child_count:2
shift state:3
process version:0, version_count:1, state:3, row:8, col:0
reduce sym:source_file_repeat1, child_count:2
shift state:3
process version:0, version_count:1, state:3, row:8, col:0
reduce sym:source_file_repeat1, child_count:2
shift state:3
process version:0, version_count:1, state:3, row:8, col:0
reduce sym:source_file_repeat1, child_count:2
```

For the example above
```
Caught #
Caught {
Caught #
Caught #
Caught #
Caught }
<=================== stalls here
```

**Reply by ahelwer** [reply]
Are you detecting whether your external scanner is called into with all symbols marked valid during error recovery?
Also try adding logic so you don't accept zero-length tokens in your external scanner.

**Reply by FoamScience** [reply]
That's exactly what it was, kept adding empty tokens :) Thanks!

**Comment by alemuller** [comment, marked as answer]
I tried to write a rule that matches anything in-between nested delimiters without external scanner.

I come up with this:

```js
module.exports = grammar({

  /* ... */

  extras: $ => [ /\s/, $.comment, $.block ],

  rules: {
    // ...

    // NOTE: Be aware of conflicts.
    //       This rule, also matches '#{' and '#}'
    comment: $ => token(/#[^\n]+/),

    block: $ => seq(
      '#{',
      repeat(choice( // allow zero repetition
        $.block,
        /([^#]|#[^{}])+/,
      )),
      '#}',
    ),

  }
)}
```

A couple of ways to resolve the conflict.

1. Change `comment`'s regex so it doesn't match tokens starting with `#{` and `#}`:

```js
comment: $ => token(/#[^{}\n][^\n]*/),

// no change to `block`
```

2. Adding precedence to `block` and `comment`:

```js
comment: $ => token(prec(1,/#[^\n]+/)), // C1

block: $ => seq(
  token(prec(2,'#{')), // B1
  repeat(choice( // allow zero repetition
    $.block,
    token(prec(2,/([^#]|#[^{}])+/)), // B2
  )),
  token(prec(2,'#}')),  // B3
),
```

Usually, when writing those "sub matching", it is common to add `token.immediate` to avoid matching other extras tokens. Here, it is
not needed because B2 should match all other extras between B1 and B2.

Some *aliasing* can be used to hide the nested blocks.

Working with token's precedence is hard, I'm might have missed something.

@ahelwer, can you think any example where this would fail?

I haven't finished editing and already though of one: `"#{}#"` (strings). But this could also be worked around with precedence.

**Reply by FoamScience** [reply]
Thanks a lot!! My comments are c-like, the only other thing which starts with "#" are pre-processor directives; but those are easy to distinguish. I confirm this works for all of my test cases; This:
```
#{
    /*
        Comment 
    */
    #include<vector>
    #include "someFile.H"

    {
        os << vector(-::sin($angle), 0, ::cos($angle));
        os << "#{ example #}";
    }
#};
```
Parses like this (which is good enough, I don't care how many code children a code node has):
```
(code (code))
```

**Reply by alemuller** [reply]
There are cases where it is not going to work. Like in strings, as I mentioned at the end of my previous post.
```
#{ printf ("#{"); #}
```

You can fix this one case with:
```js
    block: $ => seq(
      token(prec(2,'#{')),
      repeat(choice( // allow zero repetition
        $.block,
        token(prec(2,/([^#\n"]|#[^{}\n]|"([^"\\]|\\.)+")+/)),
      )),
      token(prec(2,'#}')),
    ),
```

This seems to work:

```
#{ printf ("#{"); #}
#{ printf (" \"#{ "); #}
```

The string at host language also might need some special treatment, depending how it is defined.


Again. Writing this kind of expressions isn't easy. You need to do a lot of testing to find how it can fail. 

EDIT. Just fixed a bug in the regex. It should be `\\.` instead of `\\"`.

**Reply by FoamScience** [reply]
You're right! This fails if "#{" appears unquoted inside comment in the code block too :(
But, given that I'm at early stages of development, and that case is extremely uncommon, I'm sticking with this nice RegExp for now.

**Reply by alemuller** [reply]
Yeah. I was skimming the next ISO C draft and thought about this. The regex should also ignore `#{` and `}#` when inside `''` or `<>`.

```c
#include <#{> // not illegal
const char u'{#'; // also not illegal
```

Also, `#` is a valid C operator (used inside macros). 

```c
#define foo(x) }#x // legal, i think
```

But those are corner cases. I don't think anyone in their right mind would do this.

The biggest issue are comments. But shouldn't be too hard to write a regex that matches the most common cases.
